### PR TITLE
CDEC-494: Prepare for consolidation of blund files into epoch files

### PR DESCRIPTION
**Pull request by erikd**
## Description

As per title. This is the first PR of a series of 2.

Along with the refactoring it adds a function `whenAcquireWrite` which will be used to ensure that the code to consolidate blunds into epoch files is not re-entered (ie only one thread is allow to consolidate blunds).

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-494

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: db/src/Pos/DB/Block/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    bsp <- flip getAllPaths hh . view blockDataDir <$> getNodeDBs
+    block <- getRawData $ bspBlock bsp
+    undo <- getRawData $ bspUndo bsp
+    case (,) <$> block <*> undo of
+        Just blund -> do
+            putRawData (bspBlund bsp) $ serialize' blund
+            liftIO . removeFile $ bspBlock bsp
+            liftIO . removeFile $ bspUndo bsp
+            return $ Just blund
+        Nothing -> return Nothing
```
Comment (parsonsmatt) [inline db/src/Pos/DB/Block/Internal.hs]: Might be clearer to use `Maybe`'s traversable instance:

```haskell
for ((,) <$> block <*> undo) $ \blund -> do
    -- ...
    return blund
```

File: db/src/Pos/DB/Block/Internal.hs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+data BlockStoragePaths = BlockStoragePaths
+  { bspRoot  :: FilePath
+    -- | Block data itself.
+  , bspBlock :: FilePath
+    -- | Undo information for a block.
+  , bspUndo  :: FilePath
+    -- | Combined storage format. Either this or a combination of 'Block' and
+    -- 'Undo' files will be present.
+  , bspBlund :: FilePath
+  }
```
Comment (parsonsmatt) [inline db/src/Pos/DB/Block/Internal.hs]: 4 space indent

File: util/Pos/Util/Concurrent/RWLock.hs [hunk 3]
```diff
@@ -43,3 +44,13 @@ withRead l = bracket_ (acquireRead l) (releaseRead l)
...
+
+-- | If the write lock can be acquired, perform the requested action
+-- and then release the lock. If the lock cannot be acquired, no
+-- action is performed.
+whenAcquireWrite :: (MonadIO m, MonadMask m) => RWLock -> m a -> m (Maybe a)
+whenAcquireWrite l action =
+    bracket
+        (liftIO $ RWL.tryAcquireWrite l)
+        (\ haveLock -> when haveLock $ releaseWrite l)
+        (\ haveLock -> if haveLock then Just <$> action else pure Nothing)
```
Comment (parsonsmatt) [inline util/Pos/Util/Concurrent/RWLock.hs]: the style guide says not to put a space between the `\` and the lambda argument

File: db/src/Pos/DB/Block/Internal.hs (hunk 2 above)
Comment (erikd) [reply db/src/Pos/DB/Block/Internal.hs]: You're right. That was code moved from another file :)

File: db/src/Pos/DB/Block/Internal.hs (hunk 1 above)
Comment (erikd) [reply db/src/Pos/DB/Block/Internal.hs]: Also code moved from another file, but i will fix it.

File: util/Pos/Util/Concurrent/RWLock.hs (hunk 3 above)
Comment (erikd) [reply util/Pos/Util/Concurrent/RWLock.hs]: Fixed!

**Pull request merged**